spec/unit/indirector/catalog/compiler_spec.rb

```diff
@@ -300,4 +337,114 @@ def a_request_that_contains(facts)
     end
 
   end
+
+  def build_catalog(node, num_resources, sources = nil, parameters = {:ensure => 'file'})
+    catalog = Puppet::Resource::Catalog.new(node.name, node.environment)
+
+    resources = []
+    resources << Puppet::Resource.new("notify", "alpha")
+    resources << Puppet::Resource.new("notify", "omega")
+
+    0.upto(num_resources) do |idx|
+      parameters.merge! :require => "Notify[alpha]", :before  => "Notify[omega]"
+      if sources
+        parameters.merge! :source => sources[idx % sources.size]
+      end
+      # The compiler does not operate on a RAL catalog, so we're
+      # using Puppet::Resource to produce a resource catalog.
+      agnostic_path = File.expand_path("/tmp/file_#{idx}.txt") # Windows Friendly
+      rsrc = Puppet::Resource.new("file", agnostic_path, :parameters => parameters)
+      rsrc.file = 'site.pp'
+      rsrc.line = idx+1
+      resources << rsrc
+    end
+
+    resources.each do |rsrc|
+      catalog.add_resource(rsrc)
+    end
+    catalog
+  end
+
+  describe "when inlining metadata" do
+    describe "on a catalog" do
+      let(:node) { Puppet::Node.new 'me' }
+      let(:checksum_type) { 'md5' }
+      before :each do
+        @compiler = Puppet::Resource::Catalog::Compiler.new
+        @checksum_types = Puppet.settings[:supported_checksum_types]
+      end
+
+      ['sha256.mtime.md5', 'sha256', 'md5.sha256'].each do |checksum_types|
+        describe "with agent requesting checksum_type #{checksum_types}" do
+          let(:checksum) { checksum_types.split('.').first }
+          let(:sha) {
+            sha = {'sha256' => '5891b5b522d5df086d0ff0b110fbd9d21bb4fc7163af34d08286a2e846f6be03',
+                   'md5' => 'b1946ac92492d2347c6235b4d2611184'}[checksum]
+          }
+
+          it "sets checksum and checksum_value for resources with puppet:// source URIs" do
+            catalog = build_catalog(node, 3, ['puppet:///modules/mymodule/config_file.txt'])
+            catalog.resources.select {|r| r.type == 'File'}.each do |r|
+              ral = r.to_ral
+              r.expects(:to_ral).returns(ral)
+
+              metadata = stub 'metadata'
+              metadata.stubs(:checksum).returns("{#{checksum}}#{sha}")
+              metadata.stubs(:ftype).returns("file")
+
+              source = stub 'source'
+              source.stubs(:metadata).returns(metadata)
+
+              ral.expects(:parameter).with(:source).returns(source)
```

Comment:
Stubbing this was tricky because it reaches deep into the File type. The best way to clean it up would probably be to add a `metadata` method to the File type.